core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,845 @@
+package scalaz.zio.stream
+
+import scalaz.zio._
+import scala.reflect._
+
+/**
+ * A `Chunk[A]` represents a chunk of values of type `A`. Chunks are designed
+ * are usually backed by arrays, but expose a purely functional, safe interface
+ * to the underlying elements, and they become lazy on operations that would be
+ * costly with arrays, such as repeated concatenation.
+ */
+sealed trait Chunk[@specialized +A] { self =>
+
+  /**
+   * The number of elements in the chunk.
+   */
+  def length: Int
+
+  /**
+   * Determines if the chunk is empty.
+   */
+  final def isEmpty: Boolean = length == 0
+
+  /**
+   * Determines if the chunk is not empty.
+   */
+  final def notEmpty: Boolean = length > 0
+
+  /**
+   * Returns the concatenation of this chunk with the specified chunk.
+   */
+  final def ++[A1 >: A](that: Chunk[A1]): Chunk[A1] =
+    if (self.length == 0) that
+    else if (that.length == 0) self
+    else Chunk.Concat(self, that)
+
+  /**
+   * Returns a filtered, mapped subset of the elements of this chunk.
+   */
+  def collect[@specialized B](p: PartialFunction[A, B]): Chunk[B] =
+    self.materialize.collect(p)
+
+  /**
+   * Drops the first `n` elements of the chunk.
+   */
+  final def drop(n: Int): Chunk[A] =
+    self match {
+      case Chunk.Slice(c, o, l) =>
+        Chunk.Slice(c, o + n, l - n)
+      case c @ Chunk.Singleton(_) =>
+        if (n <= 0) c else Chunk.empty
+      case _ =>
+        Chunk.Slice(self, n, self.length - n)
+    }
+
+  /**
+   * Drops all elements so long as the predicate returns true.
+   */
+  def dropWhile(f: A => Boolean): Chunk[A] = {
+    val len = self.length
+
+    var i = 0
+    while (i < len && f(self(i))) {
+      i += 1
+    }
+
+    if (i == len) Chunk.empty
+    else Chunk.Slice(self, i, len - i)
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Chunk[_] =>
+      if (self.length != that.length) false
+      else {
+        var i     = 0
+        var equal = true
+        val len   = self.length
+
+        while (equal && i < len) {
+          equal = self(i) == that(i)
```

Comment:
could do an early return here

Reply:
Would that be very different than letting the while loop end due to `equal` being false?